### PR TITLE
Raise a warning when a keyword passed to `get_output()` is unused

### PR DESCRIPTION
**Pull request by Joshua-Chin**
This resolves issue #557. `get_output()` uses introspection to retrieve the key word arguments from `get_output_for()`. All layers now have the attribute `get_output_kwargs`, which allows a user to explicitly specify the kwargs `get_output_for` will use. Also, I changed the signature of `BatchNormalizationLayer.get_output_for` to explicitly state the kwargs that it uses.

### CODE REVIEW

**Comment by benanne** [comment]
Awesome, thanks for contributing! It seems that Travis fails -- I wanted to check out why, but for some reason it's not showing me the logs right now. Can anybody else see them? I'm wondering if it's just me or if something is wrong with Travis at the moment.

**Comment by Joshua-Chin** [comment]
According to https://www.traviscistatus.com/, Travis is performing emergency database upgrades for the build log system.

**Comment by Joshua-Chin** [comment]
`mock` and `inspect` don't work together very well in python2. I am trying to find a workaround.

File: lasagne/layers/helper.py [hunk 1]
```diff
@@ -183,6 +189,13 @@ def get_output(layer_or_layers, inputs=None, **kwargs):
                                  "mapping this layer to an input expression."
                                  % layer)
             all_outputs[layer] = layer.get_output_for(layer_inputs, **kwargs)
+            if not isinstance(layer, Mock) and \
+                    not isinstance(layer.get_output_for, Mock):
```
Comment (f0k) [inline lasagne/layers/helper.py]: I'd prefer not to test for a `Mock` instance. What about `if isinstance(layer, Layer)` or `hasattr(layer, 'get_output_for')`?

File: lasagne/layers/helper.py [hunk 2]
```diff
@@ -183,6 +189,13 @@ def get_output(layer_or_layers, inputs=None, **kwargs):
...
+                unused_kwargs -= set(getargspec(layer.get_output_for)[0])
+            unused_kwargs -= set(layer.get_output_kwargs)
+    if unused_kwargs:
+        template = ", ".join(unused_kwargs)
+        warn("get_output() was called with unused kwargs: %s" % template)
```
Comment (f0k) [inline lasagne/layers/helper.py]: Could we go the other way instead -- collect a list of all accepted keywords and compare it to `set(kwargs.keys())` afterwards? This would allow us to not only print the keyword arguments that were not understood, but also the list of keyword arguments that would have been accepted (or even make suggestions using `difflib.get_close_matches`).

File: lasagne/layers/helper.py [hunk 3]
```diff
@@ -183,6 +189,13 @@ def get_output(layer_or_layers, inputs=None, **kwargs):
                                  "mapping this layer to an input expression."
                                  % layer)
             all_outputs[layer] = layer.get_output_for(layer_inputs, **kwargs)
+            if not isinstance(layer, Mock) and \
+                    not isinstance(layer.get_output_for, Mock):
+                unused_kwargs -= set(getargspec(layer.get_output_for)[0])
```
Comment (f0k) [inline lasagne/layers/helper.py]: It might be good to just collect the actual keyword arguments, i.e., the ones that have default values set:
```
names, _, _, defaults = getargspec(layer.get_output_for)
unused_kwargs -= set(names[-len(defaults):])
```

**Comment by Joshua-Chin** [comment]
I just squashed the commits into a single commit. Is there anything else I should do/fix?

File: lasagne/layers/helper.py [hunk 4]
```diff
@@ -184,6 +190,24 @@ def get_output(layer_or_layers, inputs=None, **kwargs):
...
+                    accepted_kwargs |= set(names[-len(defaults):])
+            accepted_kwargs |= set(layer.get_output_kwargs)
+    unused_kwargs = set(kwargs.keys()) - accepted_kwargs
+    if unused_kwargs:
+        suggestions = []
+        for kwarg in unused_kwargs:
+            suggestion = get_close_matches(kwarg, accepted_kwargs)
+            if suggestion:
+                suggestions.append('%s (perhaps you meant %s)'
+                                   % (kwarg, suggestion[0]))
```
Comment (f0k) [inline lasagne/layers/helper.py]: Yes, that's what I had in mind :) Very cool! :+1:

File: lasagne/layers/helper.py [hunk 5]
```diff
@@ -184,6 +190,24 @@ def get_output(layer_or_layers, inputs=None, **kwargs):
                                  "mapping this layer to an input expression."
                                  % layer)
             all_outputs[layer] = layer.get_output_for(layer_inputs, **kwargs)
+            if isinstance(layer.get_output_for, MethodType):
+                # Mock in python2 does not support introspection
```
Comment (f0k) [inline lasagne/layers/helper.py]: `layer.get_output_for` could also be a function if somebody monkey-patches an instance. This won't be a common use case, but I'm still a bit worried about making the implementation less generic just to cater for deficits in the test framework.
To make it a little more robust, what about either:
```python
if isinstance(layer.get_output_for, (MethodType, FunctionType)):
```
or:
```python
try:
    names, _, _, defaults = getargspec(layer.get_output_for)
except TypeError:
    pass  # if introspection is not possible, skip it
else:
    if defaults is not None:
        accepted_kwargs |= set(names[-len(defaults):])
```

**Comment by Joshua-Chin** [comment]
I changed the introspection code to catch `TypeError`s.

**Comment by f0k** [comment]
Great! I'm really happy with this, thanks a lot for your work! I've also ran the full test suite locally, everything works (Travis skips the GPU tests). And the `difflib` feature is excellent :)

The only open question is (@benanne, @ebenolson): Do we want to warn about the "deterministic" keyword for a network without dropouts/noise or do we want to allow "deterministic" no matter what? This could be implemented by changing [a single line](https://github.com/Lasagne/Lasagne/pull/585/files#diff-84ce884e5f993565a6a1db7b062daa20R154), so it's just a question of what we'd prefer.

An argument for always allowing "deterministic" is that it would encourage people to always pass `deterministic=True` when creating the validation/test graph, making it less likely to be forgotten. If there's a warning as long as you don't use dropout layers, people may remove it (possibly without understanding) and not add it back later. What do you think?

**Comment by ebenolson** [comment]
I'm in favor of letting `deterministic` through - I don't think you should need to modify training loop code due to changes in the network definition. But as it's only a warning I'm ok with either case.

**Comment by benanne** [comment]
I think @ebenolson's argument for letting it through is compelling: I like to keep my training loop code and actual model definitions in separate files, and I wouldn't want to have to touch the training loop code to get rid of a pesky warning, just because I happen to have a model without dropout layers :)

**Comment by Joshua-Chin** [comment]
I added `deterministic` as an accepted kwarg and squashed the commits again.

**Comment by f0k** [comment]
Perfect, thanks again for your work and for your patience! Let's finally merge this!
